diff --git a/library/venv/__init__.py b/library/venv/__init__.py
--- a/library/venv/__init__.py
+++ b/library/venv/__init__.py
@@ -422,7 +422,10 @@ def main(args=None):
                             action='store_true', dest='system_site',
                             help='Give the virtual environment access to the '
                                  'system site-packages dir.')
-        if os.name == 'nt':
+        # TODO(T61328507) symlinking does not work properly yet since
+        # sys.prefix / sys.executable is currently set after resolving all
+        # symlinks.
+        if os.name == 'nt' or sys.implementation.name == "pyro":
             use_symlinks = False
         else:
             use_symlinks = True
